core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,475 @@
+import io.grpc.Status;
+import java.net.SocketAddress;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link LoadBalancer} that provides no load-balancing over the addresses from the {@link
+ * io.grpc.NameResolver}. The channel's default behavior is used, which is walking down the address
+ * list and sticking to the first that works.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/10383")
+final class PickFirstLeafLoadBalancer extends LoadBalancer {
+  private final Helper helper;
+  private final Map<SocketAddress, SubchannelData> subchannels = new HashMap<>();
+  private Index addressIndex;
+  private volatile ConnectivityState currentState = IDLE;
+
+  PickFirstLeafLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  @Override
+  public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    List<EquivalentAddressGroup> servers = resolvedAddresses.getAddresses();
+    if (servers.isEmpty()) {
+      handleNameResolutionError(Status.UNAVAILABLE.withDescription(
+          "NameResolver returned no usable address. addrs=" + resolvedAddresses.getAddresses()
+          + ", attrs=" + resolvedAddresses.getAttributes()));
+      return false;
+    }
+    for (EquivalentAddressGroup eag : servers) {
+      if (eag == null) {
+        handleNameResolutionError(Status.UNAVAILABLE.withDescription(
+            "NameResolver returned address list with null endpoint. addrs="
+            + resolvedAddresses.getAddresses() + ", attrs=" + resolvedAddresses.getAttributes()));
+        return false;
+      }
+    }
+    // We can optionally be configured to shuffle the address list. This can help better distribute
+    // the load.
+    if (resolvedAddresses.getLoadBalancingPolicyConfig()
+        instanceof PickFirstLeafLoadBalancerConfig) {
+      PickFirstLeafLoadBalancerConfig config
+          = (PickFirstLeafLoadBalancerConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+      if (config.shuffleAddressList != null && config.shuffleAddressList) {
+        servers = new ArrayList<EquivalentAddressGroup>(servers);
+        Collections.shuffle(servers,
+            config.randomSeed != null ? new Random(config.randomSeed) : new Random());
+      }
+    }
+
+    final List<EquivalentAddressGroup> newImmutableAddressGroups =
+        Collections.unmodifiableList(new ArrayList<>(servers));
+
+    if (addressIndex == null) {
+      addressIndex = new Index(newImmutableAddressGroups);
+    } else if (currentState == READY) {
+      // If a ready subchannel exists in new address list,
+      // keep this connection and don't create new subchannels
+      SocketAddress previousAddress = addressIndex.getCurrentAddress();
```

Comment:
This is broken because current addressIndex is not pointing to the ready subchannel, right?